### ISSUE TITLE
[question] Is compiler.version necessary?

### ISSUE THREAD

**Issue by Han-sss** [conan-io/conan#14973]
### What is your question?

Hello, dear conan tool developers.
Recently our team has been using Conan 2.x to compile the c libraries, our compilation environment is as follows
- host_system = oh(an embedded operation system)
- host_arch = armv8
- compiler = clang-12.0.1
- build_system = windows 10/ ubuntu 22.04

Our problem is that we may update the compiler with the iteration of os. The compiler version will change with the upgrade, from clang-12 to clang-15.

In this case, if we compile a new library using clang-15 and depend on the clang-12 compiled library. In this case, we can't download the library directly from the repository because of the mismatch of sha256 of the libraries.

To solve this problem, we want to remove the keyword, `compiler.version` in the profile, which leads to the problem `ERROR: conanfile.py: Invalid ID: Invalid: 'settings.compiler.version' value not defined`. Could you give me some advice about this problem?

To summarize, is it possible to compile using a profile without `compiler.version`. If not, is it possible to use the library compiled using clang-12 (on the local repo) to construct the library using clang-15?

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @Han-sss 

Thanks for your question.

When you are building Conan packages, the compiler version plays an important role. Building the same library with one compiler version or another can result in quite different binaries, with different performance, resolved bugs, better security, etc. A different issue is that binaries can be compatible and you can mix binaries from different compiler versions, but the actual binaries will be different.

Conan will model each one with a different ``package_id`` for each ``compiler.version``, and in the general case it is not recommended to drop it.

Then, to implement compatibility or mixing different compiler versions you can follow different approaches:

- Explicitly define which packages use a different compiler version. Both in profiles and command line you can specify something like ``-s compiler.version=15 -s mypkg/*:compiler.version=12``, and that will use 15 for all packages except for ``mypkg`` that will use the binary compiled with clang 12
- Explicitly define general compatibility rules in the ``compatibility.py`` extension: https://docs.conan.io/2/reference/extensions/binary_compatibility.html, https://docs.conan.io/2/reference/binary_model.html#reference-binary-model

Finally, I'd like to also suggest a very simple, robust and straightforward solution, which is consider to just rebuild all the binaries with the new ``compiler.version=15``. When the CI is there, it is really not such a big effort, and it pays off, both in improvements in binaries, but also in overall simplicity for developers, tools and CI in general

**Comment by Han-sss** [comment]
> Hi @Han-sss
> 
> Thanks for your question.
> 
> When you are building Conan packages, the compiler version plays an important role. Building the same library with one compiler version or another can result in quite different binaries, with different performance, resolved bugs, better security, etc. A different issue is that binaries can be compatible and you can mix binaries from different compiler versions, but the actual binaries will be different.
> 
> Conan will model each one with a different `package_id` for each `compiler.version`, and in the general case it is not recommended to drop it.
> 
> Then, to implement compatibility or mixing different compiler versions you can follow different approaches:
> 
> * Explicitly define which packages use a different compiler version. Both in profiles and command line you can specify something like `-s compiler.version=15 -s mypkg/*:compiler.version=12`, and that will use 15 for all packages except for `mypkg` that will use the binary compiled with clang 12
> * Explicitly define general compatibility rules in the `compatibility.py` extension: https://docs.conan.io/2/reference/extensions/binary_compatibility.html, https://docs.conan.io/2/reference/binary_model.html#reference-binary-model
> 
> Finally, I'd like to also suggest a very simple, robust and straightforward solution, which is consider to just rebuild all the binaries with the new `compiler.version=15`. When the CI is there, it is really not such a big effort, and it pays off, both in improvements in binaries, but also in overall simplicity for developers, tools and CI in general

Thank you so much for your reply. 

PS: We decided to rebuild them all if necessary. ; )

**Comment by memsharded** [comment]
Sounds good, thanks for the feedback! :)
